src/Microsoft.NetCore.Analyzers/Core/Security/DoNotOverloadDeserializationBinderWithoutThrowingAnException.cs

```diff
@@ -0,0 +1,158 @@
+            DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create<DiagnosticDescriptor>(
+                DoNotOverloadSerializationBinderWithoutThrowingAnExceptionRule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    Dictionary<IMethodSymbol, CacheState> methodCache = new Dictionary<IMethodSymbol, CacheState>();
+                    WellKnownTypeProvider typeProvider = WellKnownTypeProvider.GetOrCreate(compilationStartAnalysisContext.Compilation);
+                    INamedTypeSymbol SerializationBinder = null;
+                    if (!typeProvider.TryGetTypeByMetadataName("System.Runtime.Serialization.SerializationBinder", out SerializationBinder))
+                    {
+                        System.Diagnostics.Debug.Fail("Unable to find System.Runtime.Serialization.SerializationBinder threw WellKnownTypeProvider");
+                        return;
+                    }
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        (SymbolAnalysisContext symbolAnalysisContext) =>
+                        {
+                            IMethodSymbol msym = (IMethodSymbol)symbolAnalysisContext.Symbol;
+                            if (msym.Name == "BindToType")
+                            {
+                                if (ITypeSymbolExtensions.Inherits(msym.ContainingType, SerializationBinder))
+                                {
+                                    IBlockOperation msymOps = msym.GetTopmostOperationBlock(compilationStartAnalysisContext.Compilation);
+                                    methodCache[msym] = CacheState.InProgress;
+                                    if (!hasThrow(msymOps.Operations))
+                                    {
+                                        symbolAnalysisContext.ReportDiagnostic(
+                                            Diagnostic.Create(
+                                                DoNotOverloadSerializationBinderWithoutThrowingAnExceptionRule,
+                                                msym.Locations[0],
+                                                msym.ContainingType.ToDisplayString(
+                                                    SymbolDisplayFormat.MinimallyQualifiedFormat)));
+                                    }
```

Comment:
>} [](start = 36, length = 1)

Just for completeness's sake, could have:

```csharp
else
{
    methodCache[msym] = CacheState.PresentFalse;
}
```

Reply:
Wait, who said this should be PresentFalse?  Isn't that backwards?

---
In reply to: [309484313](https://github.com/dotnet/roslyn-analyzers/pull/2733#discussion_r309484313) [](ancestors = 309484313)